### PR TITLE
Add method for creating volume label

### LINKED ISSUE #30: Changes in "util-linux" (blkid) -> LABEL is now LABEL_FATBOOT
Hey @natevw 

I just ran into a problem with software that relies on fatfs. See this [Xen Orchestra issue](https://github.com/vatesfr/xen-orchestra/issues/4449).

I am not sure if you are aware of it, but blkid > 2.33-rc1 now uses "LABEL_FATBOOT" instead of "LABEL". Is it possible to change this in fatfs, or maybe set both LABEL and LABEL_FATBOOT (to the same value)?

### PR DESCRIPTION
**Pull request by andrzejressel**
This PR adds method `createLabel` that will create special file for volume label that will be detected by `blkid` after latest changes.

I've tested it with my fork of Xen Orchestra: https://github.com/jereksel/xen-orchestra/tree/label and Ubuntu 20.04

Fixes #30

### CODE REVIEW

**Comment by natevw** [comment]
Hi sorry for the delay on this. I think this is a good start and a reasonable approach overall, but have lots of specific feedback I'll add on the PR itself soon. (About to hop on yet another meeting call…)

File: index.js [hunk 1]
```diff
@@ -557,6 +557,22 @@ exports.createFileSystem = function (volume, opts, cb) {
         else _.delayedCall(cb, S.err.NOSYS());
     }, (_n_ === '_nested_')); };
     
+    fs.createLabel = function (name, cb) {
+        fs.writeFile(name, "", () => {
+            fs.open(name, "w", (e,_fd) => {
```
Comment (natevw) [inline index.js:562]: Would it be possible to use simply one `fs.open(…, "a")` call here instead nesting of `fs.write`/`fs.open`?

File: package.json [hunk 2]
```diff
@@ -4,7 +4,11 @@
   "description": "fs implementation on top of raw FAT16/FAT32 block source",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "label-testcase": "nodejs tests/label.js"
+  },
+  "devDependencies": {
+    "promise-toolbox": "0.15.1"
```
Comment (natevw) [inline package.json:11]: See note below on usage.

File: tests/label.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+var fatfsBufferInit = require('./fatfs-buffer').init
+var fs = require('fs')
+
+const buffer = fatfsBufferInit({ label: 'cidata     ' })
+
+const { createLabel } = promisifyAll(
+  fatfs.createFileSystem(fatfsBuffer(buffer), {allowLowercaseNames: true})
+)
+
+createLabel("cidata").then(() => {
```
Comment (natevw) [inline tests/label.js:13]: I don't think it's worth bringing in a new dependency just for one call. Please match the "old school" callback style used by the rest of the project here.

File: tests/fatfs-buffer.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
+//From https://github.com/vatesfr/xen-orchestra/blob/9f0b22d3e9c4672d96cb3f63b1e0fd42609e2406/packages/xo-server/src/fatfs-buffer.js
```
Comment (natevw) [inline tests/fatfs-buffer.js:1]: Is this code restricted by the AGPL3 license then, or is its author willing to contribute it under the MIT/Apache terms of this library?

File: .gitignore [hunk 5]
```diff
@@ -1,3 +1,5 @@
 .DS_Store
 .vscode/
 node_modules/
+package-lock.json
+label.img
```
[the remark this answers is not in the document]
Comment (natevw) [reply .gitignore:5]: Oh silly me… was in a hurry and not realizing this was in the .gitignore and the whole point is that this *is* a temp file that is **not** getting tracked. Consider changing to the explicit path expected though? `tests/label.img`

File: tests/label.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+const promisifyAll = require('promise-toolbox').promisifyAll;
+var fatfs = require("../");
+var fatfsBuffer = require('./fatfs-buffer').buffer
+var fatfsBufferInit = require('./fatfs-buffer').init
+var fs = require('fs')
+
+const buffer = fatfsBufferInit({ label: 'cidata     ' })
```
Comment (natevw) [inline tests/label.js:7]: My understanding was that the buffer code from the xen-orchestra was a workaround before the `createLabel` method you're adding. So is it possible that passing in the same "cidata" label here will mask failures of the new method itself? Consider passing a recognizable "wrong" label here instead.

**Review by natevw — CHANGES_REQUESTED** [review]
Thanks much for this PR and including tests. Provided lots of feedback on some of the details but hopefully those are easy fixes and the core technique relying on `fs._updateEntry` looks good to me!

File: tests/fatfs-buffer.js (hunk 4 above)
Comment (jereksel) [reply tests/fatfs-buffer.js:1]: I think I will replace it with example driver from readme and mkfs (https://superuser.com/a/668534)

**Comment by marlluslustosa** [comment]
I tested this branch and recompiled XO locally and it didn't work.
Is that functional?

**Comment by jereksel** [comment]
> I tested this branch and recompiled XO locally and it didn't work.
> Is that functional?

Yeah, it should work fine. Could you please provide output of `blkid /dev/cdrom` and `dosfslabel /dev/cdrom`. I think you could also try to debug cloud-init. I can't remember how I've done it, but you can invoke only source search.

File: .gitignore (hunk 5 above)
[the remark this answers is not in the document]
Comment (jereksel) [reply .gitignore]: Fixed

File: index.js (hunk 1 above)
Comment (jereksel) [reply index.js]: Fixed

File: package.json (hunk 2 above)
Comment (jereksel) [reply package.json]: Fixed

File: tests/label.js (hunk 3 above)
Comment (jereksel) [reply tests/label.js]: Fixed

File: tests/fatfs-buffer.js (hunk 4 above)
Comment (jereksel) [reply tests/fatfs-buffer.js]: Fixed

File: tests/label.js (hunk 6 above)
Comment (jereksel) [reply tests/label.js]: Fixed

**Comment by jereksel** [comment]
Sorry that it took so long. I've fixed PR all comments

**Comment by marlluslustosa** [comment]
Cool. I am currently working to fix this on cloudinit. I have now sent patch patches to cloudinit both to see LABEL and LABEL_FATBOOT in their own scripts.

I hope to see this resolved by both parties as soon as possible.
You can check my update here: https://bugs.launchpad.net/cloud-init/+bug/1841466/comments/5
thanks!

File: index.js [hunk 7]
```diff
@@ -557,6 +557,20 @@ exports.createFileSystem = function (volume, opts, cb) {
         else _.delayedCall(cb, S.err.NOSYS());
     }, (_n_ === '_nested_')); };
     
+    fs.createLabel = function (name, cb) {
+        fs.open(name, "a", (e,_fd) => {
+            if (e) {
+                throw '_open_error_'
```
Comment (natevw) [inline index.js:563]: should be `return cb(e);`

**Review by natevw — CHANGES_REQUESTED** [review]
Looking good, just one hopefully final bit of feedback!

File: index.js (hunk 7 above)
Comment (jereksel) [reply index.js:563]: Is `return` required? I don't think callback has to return anything.

Comment (jereksel) [reply index.js]: And fixed

**Comment by natevw** [comment]
D'oh! Sorry to let this one linger, but thanks for cleaning it up if it's no longer needed.

**Pull request closed without merging**